### DISCUSSION TITLE
NumericUpDown read only TextBox

### DISCUSSION THREAD

**Discussion by JaoHundred** [AvaloniaUI/Avalonia#14847] [Q&A]
Hi, I'm currently trying to understand how NumericUpDown works. I want to make the TextBox part read only to prevent wrong user inputs when the field is supposed to work only with numbers. The actual IsReadOnly property affects the whole control, is there a way to apply this only to the TextBox?

**Comment by IanRawley** [comment]
Yes, via styles, but that won't limit the control to only numbers as the TextBox is what accepts the user input.  If you want to be absolutely sure it only takes numbers you'll have to intercept either KeyDown or TextInput events and filter out any inappropriate characters.

Something like this in the view Constructor: 
```
InputBox.AddHandler(TextInputEvent, InputBox_TunnelTextEvent, RoutingStrategies.Direct | RoutingStrategies.Tunnel);
```

With this implementation (which is probably not 100%)
```
private void InputBox_TunnelTextEvent(object? sender, TextInputEventArgs e)
{
    // Filter out non-digits / decimal seperators.
    e.Text = new string(e.Text?.Where((c) => Char.IsDigit(c) || c == _decimalSeperator).ToArray());
    // If there are no more characters after filtering, mark as handled so other handlers don't bother.
    if (e.Text.Equals(String.Empty)) { e.Handled = true; }
}
```

**Reply by IanRawley** [reply]
Unless what you mean is that you only want to accept input from the spinner buttons rather than allowing the user to freely type in numbers.  In that case a style like below somewhere appropriate should help:

```
<Style Selector="NumericUpDown TextBox">
	<Setter Property="IsReadOnly" Value="True"/>
</Style>
```

**Reply by JaoHundred** [reply]
Thanks for your response, yes this is the desired behavior to making only possible to change the values using the Spinner Buttons. I've tried that before too, no success for now. Im thinking if this is a similar case to the Button control where the only way of changing some properties via styles is applying inside the /template/ ContentPresenter.

**Comment by JaoHundred** [comment]
I changed the control to ButtonSpinner, you need to code some of its basic features but it did better in my case, to anyone interested here what I did.

xaml code
```
<ButtonSpinner Grid.Row="2" Margin="0,0,5,0"
               Name="buttonSpinner"
               HorizontalContentAlignment="Center"
               VerticalContentAlignment="Center"
               Spin="ButtonSpinner_Spin">

  <TextBlock Text="{Binding Person.Generations}"/>

</ButtonSpinner>
```

xaml code behind
```
private void ButtonSpinner_Spin(object? sender, SpinEventArgs e)
{
    ((PersonViewModel)DataContext!).SpinCommand.Execute(e.Direction.ToString());
}
```

ViewModel
```
[RelayCommand]
private void Spin(string direction)
{
    Person!.Generations = direction switch
    {
        "Increase" when Person!.Generations < 10 => Person!.Generations + 1,
        "Decrease" when Person!.Generations > 4 => Person!.Generations - 1,
        _ => Person!.Generations
    };
}
```

**Comment by mikhailSokolovskiy** [comment]
maybe i can rewrite main style like this 
```
        <Style Selector="NumericUpDown/template/TextBox#PART_Spinner">
            <Setter Property="IsReadOnly" Value="True"/>
        </Style>
```

**Reply by mikhailSokolovskiy** [reply]
but in my case this can't worked, i need help too)

**Reply by JaoHundred** [reply]
Hi, I could not make this work with TextBox and TextBox#Part_Spinner. I got it to work using another control which has a similar behavior to my desired case and coding its basics features, see if it can help you  https://github.com/AvaloniaUI/Avalonia/discussions/14847#discussioncomment-8721500

**Reply by mikhailSokolovskiy** [reply]
Thx, but this solution not for me, I'm sure that by all the rules of rewriting styles, everything should work fine
